File: Difficulty: Medium/Number Of Enclaves/number-of-enclaves.cpp
// User function Template for C++

#define graph_t vector<vector<int>> 
#define visit_t vector<vector<bool>>

class Solution {
  public:
    vector<int> dirx = {0,0,1,-1};
    vector<int> diry = {1,-1,0,0};
    bool valid(int i ,int j, int n, int m, graph_t& mat, visit_t& visited) {
        if(i>= 0 && i<n && j>=0 && j<m && visited[i][j]==false && mat[i][j]==1)
            return true;
        return false;
    }
    int numberOfEnclaves(vector<vector<int>> &grid) {
        // Code here
        
        int n = grid.size();
        int m = grid[0].size();
        visit_t visited(n,vector<bool>(m,false));
        queue<pair<int,int>> que;
        
        for(int i = 0;i<n;i++) {
            if(grid[i][0]==1) {
                que.push({i,0});
                visited[i][0] = true;
            }
            if(grid[i][m-1]==1) {
                que.push({i,m-1});
                visited[i][m-1] = true;
            }
        }
        
        for(int j  = 1;j<m-1;j++) {
            if(grid[0][j]==1) {
                que.push({0,j});
                visited[0][j] = true;
                
            }
            if(grid[n-1][j]==1) {
                que.push({n-1,j});
                visited[n-1][j] =true;
            }
        }
        
        while(!que.empty()) {
            pair<int,int> pr = que.front();
            que.pop();
            for(int i = 0;i<4;i++) {
                int newi = pr.first+dirx[i];
                int newj = pr.second+diry[i];
                if(valid(newi,newj,n,m,grid,visited)) {
                    que.push({newi,newj});
                    visited[newi][newj] = true;
                }
            }
        }
        
        int ans = 0;
        
        for(int i = 0;i<n;i++) {
            for(int j = 0;j<m;j++) {
                if(visited[i][j]==false && grid[i][j]==1) {
                    ans+=1;
                }
            }
        }
        
        return ans;
        
    }
};

File: Difficulty: Medium/Number Of Enclaves/number-of-enclaves.java
class Solution {
    int cntEnclaves(int[][] grid) {
        // code here
        int n = grid.length;
        int m = grid[0].length;;
        
        Deque<int[]> que = new ArrayDeque<>();
        
        for(int i = 0;i<n;i++) {
            if(i==0 || i == n-1) {
                for(int j = 0;j<m;j++) {
                    if(grid[i][j] == 1) {
                        que.add(new int[]{i,j});
                        grid[i][j] = 2;
                    }
                }
            }
            else if(grid[i][0]==1 || grid[i][m-1]==1) {
                if(grid[i][0]==1) {
                    que.add(new int[]{i,0});
                    grid[i][0] = 2;
                }
                if(grid[i][m-1]==1) {
                    que.add(new int[]{i,m-1});
                    grid[i][m-1] = 2;
                }
            }
        }
        
        
        int ans = 0;
        
        int[] dx = {0,1,0,-1};
        int[] dy = {1,0,-1,0};
        
        while(que.isEmpty()!=true) {
            int[] node = que.poll();
            for(int i = 0;i<4;i++) {
                int x = node[0] + dx[i];
                int y = node[1] + dy[i];
                if(x>=0 && x<n && y>=0 && y<m && grid[x][y]==1) {
                    grid[x][y] = 2;
                    que.add(new int[]{x,y});
                }
            }
            
        }
        
        for(int i = 0;i<n;i++) {
            for(int j = 0;j<m;j++) {
                if(grid[i][j] == 1) {
                    ans++;
                }
            }
        }
        
        return ans;
        
    }
}